src/devices/Board/Board.cs

```diff
@@ -0,0 +1,407 @@
+
+        public virtual void ReleasePin(int pinNumber, PinUsage usage, object owner)
+        {
+            if (!_initialized)
+            {
+                throw new InvalidOperationException("Cannot release a pin if board is not initialized.");
+            }
+
+            int logicalPin = ConvertPinNumber(pinNumber, DefaultPinNumberingScheme, PinNumberingScheme.Logical);
+
+            lock (_pinReservationsLock)
+            {
+                if (_pinReservations.TryGetValue(logicalPin, out List<PinReservation> reservations))
+                {
+                    if (!UsageCanSharePins(usage))
+                    {
+                        PinReservation reservation = reservations.Single();
+                        if (reservation.Owner != owner || reservation.Usage != usage)
+                        {
+                            throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because you are not the owner or the usage is wrong. Class {reservation.Owner} has reserved the Pin for {reservation.Usage}");
+                        }
+
+                        _pinReservations.Remove(logicalPin);
+                    }
+                    else
+                    {
+                        PinReservation reservation = reservations.FirstOrDefault(x => x.Owner == owner);
+                        if (reservation == null)
+                        {
+                            throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because you are not a valid owner.");
+                        }
+
+                        if (reservation.Usage != usage)
+                        {
+                            throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because you are not the owner or the usage is wrong. Class {reservation.Owner} has reserved the Pin for {reservation.Usage}");
+                        }
+
+                        reservations.Remove(reservation);
+                        if (reservations.Count == 0)
+                        {
+                            _pinReservations.Remove(logicalPin);
+                        }
+                    }
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because it is not reserved.");
+                }
+            }
+        }
+
+        public abstract PinUsage DetermineCurrentPinUsage(int pinNumber);
+
+        /// <summary>
+        /// Override this method if something special needs to be done to use the pin for the given device.
+        /// Many devices support multiple functions per Pin, but not at the same time, so that some kind of
+        /// multiplexer needs to be set accordingly.
+        /// </summary>
+        /// <param name="pinNumber">The logical pin number to use.</param>
+        /// <param name="usage">The intended usage</param>
+        protected virtual void ActivatePinMode(int pinNumber, PinUsage usage)
```

Comment:
if anything this should be throwing by default but I'd rather keep this concept out of the initial abstraction

Reply:
This also belongs to the main feature: It allows (see the RaspberryPiBoard specific implementation) to set the alternate pin mode for a pin. This internally calls SetAlternatePinMode on the Raspi